### PR TITLE
 pretalx: init 2023.1.3 

### PR DESCRIPTION
**Pull request by SuperSandro2000**
## Description of changes

Deployed it for 20 years Datenspuren.

based on #231684

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/services/web-apps/pretalx.nix [hunk 1]
```diff
@@ -0,0 +1,414 @@
...
+        "--name=pretalx"
+        "--workers=4"
+        "--max-requests=1200"
+        "--max-requests-jitter=50"
+        "--log-level=info"
+      ];
+      description = lib.mdDoc ''
+        Extra arguments to pass to gunicorn.
+      '';
+      apply = lib.escapeShellArgs;
```
Comment (pbsds) [inline nixos/modules/services/web-apps/pretalx.nix:62]: TIL :+1:

File: pkgs/servers/web-apps/pretalx/default.nix [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+    "django-filter"
+    "django-formtools"
+    "libsass"
+    "markdown"
+    "pillow"
+  ];
+
+  propagatedBuildInputs = with python.pkgs; [
+    beautifulsoup4
+    beautifulsoup4.optional-dependencies.lxml
```
Comment (pbsds) [inline pkgs/servers/web-apps/pretalx/default.nix:100]: Isn't this a list? 
```suggestion
    ] ++ beautifulsoup4.optional-dependencies.lxml ++ [ 
```

Or are nested lists fine?

File: pkgs/servers/web-apps/pretalx/default.nix [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    redis = with python.pkgs; [
+      redis
+    ];
+  };
+
+  postBuild = ''
+    rm -r ./src/pretalx/frontend/schedule-editor
+    ln -s ${frontend}/lib/node_modules/@pretalx/schedule-editor ./src/pretalx/frontend/schedule-editor
+
+    PYTHONPATH=$PYTHONPATH:./src python -m pretalx rebuild
```
Comment (pbsds) [inline pkgs/servers/web-apps/pretalx/default.nix:152]: This line could use a comment

**Review by pbsds — COMMENTED** [review]
Result of `nixpkgs-review pr 280821` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
    <li>tests.nixos-functions.nixos-test</li>
  </ul>
</details>
<details>
  <summary>3 packages built:</summary>
  <ul>
    <li>pretalx</li>
    <li>pretalx.dist</li>
    <li>pretalx.static</li>
  </ul>
</details>

The package definition and build result looks fine, nixos test runs fine, and the manual builds fine and LGTM.

It seems the pytest check `test_change_custom_domain` fails on darwin, and the CI check for by-name will likely trigger on a force push.

Some nits and comments on my end, some of which may not be actionable.

File: nixos/modules/services/web-apps/pretalx.nix (hunk 1 above)
Comment (mweinelt) [reply nixos/modules/services/web-apps/pretalx.nix:62]: You're welcome.

File: nixos/modules/services/web-apps/pretalx.nix [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+        recommendedTlsSettings = lib.mkDefault true;
+        upstreams.pretalx.servers."unix:/run/pretalx/pretalx.sock" = { };
+        virtualHosts.${cfg.nginx.domain} = {
+          # https://docs.pretalx.org/administrator/installation.html#step-7-ssl
+          extraConfig = ''
+            more_set_headers Referrer-Policy same-origin;
+            more_set_headers X-Content-Type-Options nosniff;
+          '';
+          locations = {
+            "/".proxyPass = "http://pretalx";
```
[the remark this answers is not in the document]
Comment (mweinelt) [reply nixos/modules/services/web-apps/pretalx.nix:292]: Refers to an nginx upstream definition (see L284).

File: pkgs/servers/web-apps/pretalx/default.nix (hunk 2 above)
Comment (mweinelt) [reply pkgs/servers/web-apps/pretalx/default.nix:100]: No, should be concat.

File: pkgs/servers/web-apps/pretalx/default.nix (hunk 3 above)
Comment (mweinelt) [reply pkgs/servers/web-apps/pretalx/default.nix:152]:
```suggestion
    # Generate all static files; https://docs.pretalx.org/administrator/commands.html#python-m-pretalx-rebuild
    PYTHONPATH=$PYTHONPATH:./src python -m pretalx rebuild
```

File: nixos/modules/services/web-apps/pretalx.nix (hunk 4 above)
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply nixos/modules/services/web-apps/pretalx.nix:292]: That's standard nginx syntax on how to refer to upstreams. I don't think we need a comment here.

File: nixos/modules/services/web-apps/pretalx.nix [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+      '')
+    ];
+
+    services = {
+      nginx = lib.mkIf cfg.nginx.enable {
+        enable = true;
+        recommendedGzipSettings = lib.mkDefault true;
+        recommendedOptimisation = lib.mkDefault true;
+        recommendedProxySettings = lib.mkDefault true;
+        recommendedTlsSettings = lib.mkDefault true;
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply nixos/modules/services/web-apps/pretalx.nix:283]: This option only exists globally, so we can't set it local.

File: nixos/modules/services/web-apps/pretalx.nix (hunk 1 above)
Comment (SuperSandro2000) [reply nixos/modules/services/web-apps/pretalx.nix:62]: You probably mean ``utils.escapeSystemdExecArgs``.

File: nixos/modules/services/web-apps/pretalx.nix [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+  configFile = format.generate "pretalx.cfg" cfg.settings;
+
+  extras = cfg.package.optional-dependencies.redis
+    ++ lib.optionals (cfg.settings.database.backend == "mysql") cfg.package.optional-dependencies.mysql
+    ++ lib.optionals (cfg.settings.database.backend == "postgresql") cfg.package.optional-dependencies.postgres;
+
+  pythonEnv = cfg.package.python.buildEnv.override {
+    extraLibs = [ (cfg.package.python.pkgs.toPythonModule cfg.package) ]
+      ++ (with cfg.package.python.pkgs; [ gunicorn ]
+      ++ lib.optional cfg.celery.enable celery) ++ extras;
```
[the remark this answers is not in the document]
Comment (SuperSandro2000) [reply nixos/modules/services/web-apps/pretalx.nix:20]: It should only rebuild the env and not trigger rebuilds of dependency packages.

**Comment by SuperSandro2000** [comment]
> It seems the pytest check `test_change_custom_domain` fails on darwin, and the CI check for by-name will likely trigger on a force push.

Upstream doesn't mention darwin/macos in their docs and I don't want to spend my time maintaining that, mostly because I don't think anyone will host this on darwin, hence I am just setting platforms to linux.

**Review by pbsds — APPROVED** [review]

**Comment by mweinelt** [comment]
Missing a release notes entry.

**Comment by pbsds** [comment]
* #285326

**Pull request merged**